From 2639d644ab1cccca5b9fa12f63400fba3e87d2af Mon Sep 17 00:00:00 2001
From: Exverge <exverge@exverge.xyz>
Date: Sat, 20 Jun 2026 16:07:43 -0400
Subject: [PATCH 4/4] interface: Boost should not be in public interface

Realistically using an array is probably marginally faster than
using a runtime-allocated one so I'd consider this an optimization

There's probably a cleaner way to initialize the array but I'm lazy

diff --git a/src/dynarmic/backend/arm64/exclusive_monitor.cpp b/src/dynarmic/backend/arm64/exclusive_monitor.cpp
index 5c16988f..ca627c8c 100644
--- a/src/dynarmic/backend/arm64/exclusive_monitor.cpp
+++ b/src/dynarmic/backend/arm64/exclusive_monitor.cpp
@@ -15,7 +15,7 @@
 namespace Dynarmic {
 
 ExclusiveMonitor::ExclusiveMonitor(std::size_t processor_count)
-        : exclusive_addresses(processor_count, INVALID_EXCLUSIVE_ADDRESS), exclusive_values(processor_count) {}
+        : exclusive_addresses{INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS}, exclusive_values{0, 0, 0, 0} {}
 
 size_t ExclusiveMonitor::GetProcessorCount() const {
     return exclusive_addresses.size();
diff --git a/src/dynarmic/backend/loongarch64/exclusive_monitor.cpp b/src/dynarmic/backend/loongarch64/exclusive_monitor.cpp
index 24aa4853..4286ab23 100644
--- a/src/dynarmic/backend/loongarch64/exclusive_monitor.cpp
+++ b/src/dynarmic/backend/loongarch64/exclusive_monitor.cpp
@@ -8,7 +8,7 @@
 namespace Dynarmic {
 
 ExclusiveMonitor::ExclusiveMonitor(std::size_t processor_count)
-        : exclusive_addresses(processor_count, INVALID_EXCLUSIVE_ADDRESS), exclusive_values(processor_count) {}
+        : exclusive_addresses{INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS}, exclusive_values{0, 0, 0, 0} {}
 
 size_t ExclusiveMonitor::GetProcessorCount() const {
     return exclusive_addresses.size();
diff --git a/src/dynarmic/backend/riscv64/exclusive_monitor.cpp b/src/dynarmic/backend/riscv64/exclusive_monitor.cpp
index b3585c64..062f953b 100644
--- a/src/dynarmic/backend/riscv64/exclusive_monitor.cpp
+++ b/src/dynarmic/backend/riscv64/exclusive_monitor.cpp
@@ -8,7 +8,7 @@
 namespace Dynarmic {
 
 ExclusiveMonitor::ExclusiveMonitor(std::size_t processor_count)
-        : exclusive_addresses(processor_count, INVALID_EXCLUSIVE_ADDRESS), exclusive_values(processor_count) {}
+        : exclusive_addresses{INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS}, exclusive_values{0, 0, 0, 0} {}
 
 size_t ExclusiveMonitor::GetProcessorCount() const {
     return exclusive_addresses.size();
diff --git a/src/dynarmic/backend/x64/exclusive_monitor.cpp b/src/dynarmic/backend/x64/exclusive_monitor.cpp
index 0ae3e2ef..23065e8c 100644
--- a/src/dynarmic/backend/x64/exclusive_monitor.cpp
+++ b/src/dynarmic/backend/x64/exclusive_monitor.cpp
@@ -15,7 +15,7 @@
 namespace Dynarmic {
 
 ExclusiveMonitor::ExclusiveMonitor(std::size_t processor_count)
-        : exclusive_addresses(processor_count, INVALID_EXCLUSIVE_ADDRESS), exclusive_values(processor_count) {}
+        : exclusive_addresses{INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS, INVALID_EXCLUSIVE_ADDRESS}, exclusive_values{0, 0, 0, 0} {}
 
 size_t ExclusiveMonitor::GetProcessorCount() const {
     return exclusive_addresses.size();
diff --git a/src/dynarmic/interface/exclusive_monitor.h b/src/dynarmic/interface/exclusive_monitor.h
index 566743c7..0f0a0206 100644
--- a/src/dynarmic/interface/exclusive_monitor.h
+++ b/src/dynarmic/interface/exclusive_monitor.h
@@ -9,7 +9,6 @@
 #include <cstddef>
 #include <cstdint>
 #include <cstring>
-#include <boost/container/static_vector.hpp>
 
 #include <dynarmic/common/spin_lock.h>
 
@@ -80,8 +79,8 @@ private:
     static constexpr VAddr RESERVATION_GRANULE_MASK = 0xFFFF'FFFF'FFFF'FFFFull;
     static constexpr VAddr INVALID_EXCLUSIVE_ADDRESS = 0xDEAD'DEAD'DEAD'DEADull;
     static constexpr size_t MAX_NUM_CPU_CORES = 4; // Sync with src/core/hardware_properties
-    boost::container::static_vector<VAddr, MAX_NUM_CPU_CORES> exclusive_addresses;
-    boost::container::static_vector<Vector, MAX_NUM_CPU_CORES> exclusive_values;
+    std::array<VAddr, MAX_NUM_CPU_CORES> exclusive_addresses;
+    std::array<Vector, MAX_NUM_CPU_CORES> exclusive_values;
     SpinLock lock;
 };
 
-- 
2.54.0

